app/assets/javascripts/admin/models/flagged-post.js.es6

```diff
@@ -62,11 +62,74 @@ export default Post.extend({
   },
 
   deferFlags(deletePost) {
-    return ajax('/admin/flags/defer/' + this.id, { type: 'POST', cache: false, data: { delete_post: deletePost } }).catch(popupAjaxError);
+    const action = () => {
+      return ajax('/admin/flags/defer/' + this.id, {
+        type: 'POST', cache: false, data: { delete_post: deletePost }
+      });
+    };
+
+    if (deletePost && this._hasDeletableReplies()) {
+      return this._actOnFlagAndDeleteReplies(action);
+    } else {
+      return action().catch(popupAjaxError);
+    }
   },
 
   agreeFlags(actionOnPost) {
-    return ajax('/admin/flags/agree/' + this.id, { type: 'POST', cache: false, data: { action_on_post: actionOnPost } }).catch(popupAjaxError);
+    const action = () => {
+      return ajax('/admin/flags/agree/' + this.id, {
+        type: 'POST', cache: false, data: { action_on_post: actionOnPost }
+      });
+    };
+
+    if (actionOnPost === 'delete' && this._hasDeletableReplies()) {
+      return this._actOnFlagAndDeleteReplies(action);
+    } else {
+      return action().catch(popupAjaxError);
+    }
+  },
+
+  _hasDeletableReplies() {
+    return this.get('post_number') > 1 && this.get('reply_count') > 0;
+  },
+
+  _actOnFlagAndDeleteReplies(action) {
+    return new Ember.RSVP.Promise((resolve, reject) => {
```

Comment:
Not sure if I'm missing something but why do we need to return a promise here? We don't seem to be using the promise anywhere.

Reply:
It's waiting on the promise in the `removeAfter` calls: https://github.com/discourse/discourse/blob/ce97759418255af409693b95ca23f1274df8cd51/app/assets/javascripts/select-kit/components/admin-delete-flag-dropdown.js.es6#L68-L76